src/plugins/roam/roam.js

```diff
@@ -0,0 +1,254 @@
+        // and apply changes to dataset of present step. Ready to update "canvas".
+        function roamMove( ...direction ) {
+            var vector = getVectorQuaternion( ...direction );
+            presentStep.dataset.x = ( parseFloat( presentStep.dataset.x ) || 0 ) + vector[ 1 ] * 500 / spf;
+            presentStep.dataset.y = ( parseFloat( presentStep.dataset.y ) || 0 ) + vector[ 2 ] * 500 / spf;
+            presentStep.dataset.z = ( parseFloat( presentStep.dataset.z ) || 0 ) + vector[ 3 ] * 500 / spf;
+        }
+
+        // Respond every roam-controlling key one after another.
+        Array.from( roams.keys ).forEach( ( key ) => {
+            if ( roams[ key ].start === -1 ) {
+                roams[ key ].start = timestamp;
+            }
+            switch ( key ) {
+                // Moving part.
+                case "w": // forward
+                    roamMove( 0, 0, -1 );
+                    break;
+                case "s": // back
+                    roamMove( 0, -0, 1 );
+                    break;
+                case "a": // left
+                    roamMove( -1, 0, 0 );
+                    break;
+                case "d": // right
+                    roamMove( 1, 0, 0 );
+                    break;
+                case "q": // up
+                    roamMove( 0, -1, 0 );
+                    break;
+                case "e": // down
+                    roamMove( 0, 1, 0 );
+                    break;
+
+                // Rotating part.
+                case "i": // pitch-up
+                    presentStep.dataset.rotateX = ( parseFloat( presentStep.dataset.rotateX ) || 0 ) + 30 / spf;
+                    break;
+                case "k": // pitch-down
+                    presentStep.dataset.rotateX = ( parseFloat( presentStep.dataset.rotateX ) || 0 ) - 30 / spf;
+                    break;
+                case "j": // roll-left
+                    presentStep.dataset.rotateZ = ( parseFloat( presentStep.dataset.rotateZ ) || 0 ) - 30 / spf;
+                    break;
+                case "l": // roll-right
+                    presentStep.dataset.rotateZ = ( parseFloat( presentStep.dataset.rotateZ ) || 0 ) + 30 / spf;
+                    break;
+                case "u": // yaw-left
+                    presentStep.dataset.rotateY = ( parseFloat( presentStep.dataset.rotateY ) || 0 ) + 30 / spf;
+                    break;
+                case "o": // yaw-right
+                    presentStep.dataset.rotateY = ( parseFloat( presentStep.dataset.rotateY ) || 0 ) - 30 / spf;
+                    break;
+                default:
+                    break;
+            }
+        } );
+
+        // Everything's done.
+        // Just leave the impress to update the "canvas" and we've finished the work for one frame!
+        // As there no more frame between present one and the comming one, duration should be 0.
+        api.goto( presentStep.id, 0 );
+    }
+
+    function eventHandler( event ) {
```

Comment:
Please make sure this plugin allows typing text into form elements. See https://github.com/impress/impress.js/blob/master/src/plugins/form/form.js